### DISCUSSION TITLE
Contributing to tsai

### DISCUSSION THREAD

**Discussion by dnth** [timeseriesAI/tsai#130] [General]
I am fairly new to fastai and `tsai`. However I am inspired to start contributing to `tsai` because I think it really helps make time series problem more accessible. `tsai` has also helped me tremendously in my time series regression problem. 

Is learning about `nv_dev` the first step to start contributing? Or would you point me to other resources?
Is there any other aspect of `tsai` that anyone can help to improve?

Thanks for the hard work in maintaining this awesome lib.

**Comment by oguiza** [comment]
That’s great @dnth!
Thanks for offering to contribute to tsai. 
Learning nbdev would be a good start. Something I used to learn more about it is to create a dummy repo, clone it to my laptop, make changes and commit them. It’s easier to practice when you know you can’t break anything. 
I’d say there are 4 requirements where some expertise is required:

- Git & GitHub 
- nbdev
- fastai
- tsai: it’d be good to understand the tutorial notebooks.

There are multiple ways in which tsai can be improved/ developed. I always have more ideas than time, and I’m you’ll bring your as well. So if you are interested, there’s lot of areas where you can contribute.

**Reply by dnth** [reply]
Sign me up!

**Comment by oguiza** [comment]
Hi @dnth,
I'd like to know more about what areas of `tsai` you are interested in developing. In my view, there are many :)
Here's a list of potential areas I'd like to integrate into the library at some point in the future: 

* tasks: 
    * forecasting
    * imputation
    * anomaly detection
    * segmentation
    * clustering
    * self-supervised
    * semi-supervised
* type of time series: 
   * missing values
   * irregular
   * with covariates (categorical and continuous metadata)
* types of models:
    *  transformers
    * multimodal (time series + categorical + continuous)
    * tabular models
    * new state-of-the-art models: CNNs, Transformers, Hybrid, ...
* Interpretability:
    * Integration of Captum
* MLOps: 
    * Docker
    * TorchServe
* maintenance:
    * fix bugs
    * improve testing
    * improve documentation
    * adapt the library to fastai v3 when available
    * adapt `tsai` to package upgrades (Pytorch, ...)

As you can see, there're a lot of things that could be done. That's why I'm more than happy to accept contributors to `tsai`. 

Are you interested in any of these in particular?

**Reply by vrodriguezf** [reply]
wow that is an awesome todo list Ignacio!!! :)

**Reply by oguiza** [reply]
Yeah, @vrodriguezf, but I'm afraid I won't be able to complete it by Monday 🤣
This is more meant to be a list of topics I'd like to cover in the future. Topics not included in the list will have lower priority.

**Reply by dnth** [reply]
Wow that is a big list! I like to learn and contribute more in types of models. Recently, im quite fascinated with using reinforcement learning with time series. Any chance that makes into the list?

**Reply by oguiza** [reply]
Could you please share an example (maybe a paper) of this (RL applied to TS)?

**Reply by williamsdoug** [reply]
Hi @dnth.  Here's a couple startup issues I encountered when submitting my first pull request to tsai:

- After you clone the tsai repo, be sure to execute the command ```nbdev_install_git_hooks```.  This needs to be done only once, but must to be performed before any new commits.
- Before executing ```nbdev_build_docs``` all package dependencies must be available.  In my case since I use Conda, I execute ```conda activate <your_env_name>```.  At build time nbdev seems to ignore the kernel associated with your jupyter notebooks, hence the need to explicitly specify environment.

**Reply by dnth** [reply]
@williamsdoug thanks for the heads up! Those two points could potentially save me hours :)

@oguiza i realized RL applied to TS mostly are in the case of algorithmic trading time series.. Probably out of TSAI scope..

**Reply by oguiza** [reply]
Yes, Reinforcement Learning is out of the scope of this library for now. 
I agree it's a very interesting area but as you said, there's still a very long list of pending items that I believe may add substantial value to practitioners working with time-series data.
